bpo-43757: make pathlib use os.path.realpath() to resolve all symlinks in a path

barneygale opened the pull request:
Removes a pathlib-specific implementation of `realpath()` that's mostly
copied from `ntpath` and `posixpath`.


<!-- issue-number: [bpo-43757](https://bugs.python.org/issue43757) -->
https://bugs.python.org/issue43757
<!-- /issue-number -->

eryksun commented on Lib/pathlib.py:
```diff
@@ -463,6 +391,8 @@ def group(self, path):
 
     getcwd = os.getcwd
 
+    realpath = staticmethod(os.path.realpath)
```
`os.path.realpath()` is always non-strict. That's why the implementation is duplicated in pathlib. To use the `os.path` implementation, it needs a `strict=False` keyword-only option.

barneygale replied:
Is checking the return value not sufficient? Trying to avoid touching `os.path` stuff if I can.

eryksun replied:
Maybe that suffices. It could be a symlink to a path that doesn't exist, so you need to check `stat()` to ensure that it exists. That's needed even in non-strict mode, because pathlib always raises for a symlink loop. In POSIX, a symlink loop should fail with `errno` `ELOOP`. I suppose all other errors should be ignored in non-strict mode.

In Windows, a name-surrogate loop (symlink, mountpoint, etc) will fail with `winerror` `ERROR_CANT_RESOLVE_FILENAME` (1921). This will definitely change the behavior of the current implementation, in terms of errors that get ignored and paths that can be partially resolved, since `ntpath.realpath()` is much more fully developed.

barneygale replied:
Really good info, thanks.

What do you think the best course of action would be? I'm happy to add a 'strict mode' to `realpath()` if that looks like the best way forward. There's two things that can be controlled:

- Whether an exception is raised if we encounter a symlink loop (`raise_on_loop=False`?)
- Whether we expect the entire path to exist (`raise_on_missing=False`?)

barneygale replied:
In fact, maybe those two are essentially the same thing. The question is whether we proceed with a "best effort" resolution of the path by tacking on the remaining path components, or give up and raise an exception if we hit an error (not found, symlink loop, etc). That can fit under a single `strict` keyword-only argument to `realpath`. Updating patch incoming...

barneygale replied:
OK, new patch for your consideration. This one adjusts `realpath()` as follows:

- Always raise an `OSError(errno.ELOOP, ...)` when a symlink loop is detected, rather than returning a path that still includes symlinks. Rationale why this behaviour change is OK:
  - The previous behaviour makes no sense and just kicks the symlink-loop issue down the road
  - POSIX `realpath` is documented to return ELOOP when it encounters a symlink loop
  - Surely no one can be relying on this!
- Add a keyword-only `strict` argument that raises an `OSError` if the path (or an ancestor) is inaccessible/missing when `True`.

barneygale committed:
bpo-43757: make pathlib use os.path.realpath() to resolve all symlinks in a path

Removes a pathlib-specific implementation of `realpath()` that's mostly
copied from `ntpath` and `posixpath`.

barneygale committed:
Adjust `os.path.realpath` to match `pathlib.Path.resolve()` behaviour when a new `strict=True` keyword-only argument is passed.

barneygale committed:
Call `os.stat()` to raise `OSError(errno.ELOOP, ...)`.

barneygale committed:
Remove unused import

eryksun commented on Lib/pathlib.py:
```diff
@@ -1134,15 +1060,13 @@ def resolve(self, strict=False):
+        try:
+            p = self._accessor.realpath(self, strict=strict)
+        except OSError as ex:
+            if ex.errno == ELOOP:
```
The behavior in Windows could be changed to raise `RuntimeError` for `ERROR_CANT_RESOLVE_FILENAME` (1921), which would agree with the documentation. For example: `ex.errno == ELOOP or getattr(ex, 'winerror', 0) == 1921`.

It would be simpler if `winerror_to_errno()` in "PC/errmap.h" mapped `ERROR_CANT_RESOLVE_FILENAME` to `ELOOP`. But I'm not certain that it's always used equivalently to `ELOOP`. The only NT status code that maps to this Windows error code is `STATUS_REPARSE_POINT_NOT_RESOLVED`, which is due to exhausting the maximum allowed number of reparse attempts when opening a path (currently 63). But a system component could directly use `ERROR_CANT_RESOLVE_FILENAME` in a broader sense, in which case mapping it to `ELOOP` would be misleading.

barneygale committed:
Stop ignoring ERROR_CANT_RESOLVE_FILENAME error in _getfinalpathname_nonstrict()

barneygale committed:
Fix symlink detection and `posixpath` tests.

barneygale committed:
Raise RuntimeError in Path.resolve() if realpath() raises ERROR_CANT_RESOLVE_FILENAME

barneygale committed:
Remove unused import

barneygale committed:
Make documentation a little more general to accurately cover Windows.

barneygale committed:
First pass on fixing `ntpath` tests

barneygale committed:
Second pass on fixing `ntpath` tests

barneygale commented on Lib/test/test_ntpath.py:
```diff
@@ -369,19 +377,18 @@ def test_realpath_symlink_loops(self):
         self.assertPathEqual(ntpath.realpath(ABSTFN + "1\\..\\"
                                              + ntpath.basename(ABSTFN) + "y"),
                              ABSTFN + "x")
```
I'm a bit surprised about these `assertPathEqual` asserts passing - looks like `_getfinalpathname()` normalizes before it resolves symlinks, whereas (edit: posix) `readlink()` does not.

eryksun replied:
`nt._getfinalpathname()` opens a handle to a file/directory with `CreateFileW()` and calls `GetFinalPathNameByHandleW()`. The latter makes a few system calls to get the final opened path in the filesystem (e.g. "\\Windows\\explorer.exe") and the canonical DOS name of the volume device on which the filesystem is mounted (e.g. "\\Device\\HarddiskVolume2" -> "\\\\?\\C:") in order to return a canonical DOS path (e.g. "\\\\?\\C:\\Windows\\explorer.exe").

Opening a handle with `CreateFileW()` entails first getting a fully-qualified and normalized NT path, which, among other things, entails resolving ".." components naively in the path string. This does not take reparse points such as symlinks and mountpoints into account. The only time Windows parses ".." components in an opened path the way POSIX does is in the kernel when they're in the target path of a relative symlink.

`nt.readlink()` opens a handle to the file with the flag `FILE_FLAG_OPEN_REPARSE_POINT`. If the final path component is a reparse point, it opens it instead of traversing it. Then the reparse point is read with the filesystem control request, `FSCTL_GET_REPARSE_POINT`. System symlinks and mountpoints (`IO_REPARSE_TAG_SYMLINK` and `IO_REPARSE_TAG_MOUNT_POINT`) are the only supported name-surrogate reparse-point types, though `os.stat()` and `os.lstat()` handle all name-surrogate types as 'links'. Moreover, only symlinks get the `S_IFLNK` mode flag in a stat result, because they're the only ones we can create with `os.symlink()` to satisfy the usage `if os.path.islink(src): os.symlink(os.readlink(src), dst)`.

barneygale replied:
I think I follow! What would it take to do a POSIX-style "normalize as we resolve", and would we want to? I guess we'd need to call `nt._getfinalpathname()` on each path component in turn (`C:\`, `C:\Users`, `C:\Users\Barney` etc), which from my pretty basic Windows knowledge might be rather slow if that involves file handles.

Perhaps another option would be to add a note to the `os.path.realpath()` documentation explaining the different normalization/resolution order across Windows and posix

Or perhaps we just ignore it, as a longstanding and esoteric issue? :)

zooba replied:
I'd add a note. Dangling symlinks or junctions ought to be rare, and in any case if you're resolving them then you ought to be prepared for the fact that you're doing something that's likely unnecessary, undesirable, and unreliable.

("Unnecessary and undesirable" may seem a bit strong, but the whole point is that you don't need to resolve a path to do 99% of actions, because the OS will do it. So resolve() is a fairly advanced function that I think is allowed to be closer to the OS behaviour than strictly consistent. (It might be worth adding this kind of note to the docs...))

eryksun replied:
> What would it take to do a POSIX-style "normalize as we resolve",
> and would we want to? I guess we'd need to call nt._getfinalpathname()
> on each path component in turn (C:\, C:\Users, C:\Users\Barney etc),
> which from my pretty basic Windows knowledge might be rather slow if
> that involves file handles.

You asked, so I decided to write up an outline of what implementing a POSIX-style `realpath()` might look like in Windows. At its core, it's similar to POSIX: lstat(), and, for a symlink, readlink() and recur. The equivalent calls in Windows are the following:

* `CreateFileW()` (open a handle)
* `GetFileInformationByHandleEx()`: `FileAttributeTagInfo`
* `DeviceIoControl()`: `FSCTL_GET_REPARSE_POINT`

A symlink has the reparse tag `IO_REPARSE_TAG_SYMLINK`.

Filesystem mountpoints (aka junctions, which are like Unix bind mountpoints) must be retained in the resolved path in order to correctly resolve relative symlinks such as "\\spam" (relative to the resolved device) and "..\\..\\spam". Anyway, this is consistent with the UNC case, since mountpoints on a remote server can never be resolved (i.e. a final UNC path never resolves mountpoints).

Here are some of the notable differences compared to POSIX:

* If the source path is not a "\\\\?\\" verbatim path, `GetFullPathNameW()` must be called initially.  However, ".." components in the target path of a relative symlink must be resolved the POSIX way, else symlinks in the target path may be removed incorrectly before their target is resolved (e.g. "foo\\symlink\\..\\bar" incorrectly resolved as "foo\\bar"). The opened path is initially normalized as follows: 
    * replace forward slashes with backslashes
    * collapse repeated backslashes (except the UNC root must have exactly two backslashes)
    * resolve a relative path (e.g. "spam"), drive-relative path (e.g. "Z:spam"), or rooted path (e.g. "\\spam") as a fully-qualified path (e.g. "Z:\\eggs\\spam")
    * resolve "." and ".." components in the opened path (naive to symlinks)
    * strip trailing spaces and dots from the final component (e.g. "C:\\spam. . ." -> "C:\\spam")
    * resolve reserved device names in the final component of a non-UNC path (e.g. "C:\\nul" -> "\\\\.\\nul")
* Substitute drives (e.g. created by "subst.exe", or `DefineDosDeviceW`) and mapped drives (e.g. created by "net.exe", or `WNetAddConnection2W`) must be resolved, respectively via `QueryDosDeviceW()` and `WNetGetUniversalNameW()`. Like all DOS 'devices', these drives are implemented as object symlinks (i.e. symlinks in the object namespace, not to be confused with filesystem symlinks). The target path of these drives, however, is not a Device object, but rather a filesystem path on a device that can include any number of path components, some of which may be filesystem symlinks that need to be resolved. Normally when a path is opened, the system object manager reparses all DOS 'devices' to the path of an actual Device object, or a path on a Device object, before the I/O manager's parse routine ever sees the path. Such drives need to be resolved whenever parsing starts or restarts at a drive, but the result can be cached in case multiple filesystem symlinks target the same drive. 
    * Substitute drives can target paths on other substitute drives, so `QueryDosDeviceW()` has to be called in a loop that accumulates the tail path components until it reaches a real device (i.e. a target path that doesn't begin with "\\??\\"). 
    * `WNetGetUniversalNameW()` has to be called after resolving substitute drives. It resolves the underlying UNC  path of a mapped drive. The target path of the object symlink that implements a mapped drive is of the form "\\Device\\\<redirector device name\>\\;\<something\>\\server\\share\\some\\filesystem\\path". The "redirector device name" component is usually (post Windows Vista) an object symlink to a path on the system's Multiple UNC Provider (MUP) device, "\\Device\\Mup". The mapped-drive target path ultimately resolves to a redirected filesystem that's mounted in the MUP device namespace at the "share" name. This is an implementation detail of the filesystem redirector and MUP device, which the Multiple Provider Router (MPR) WNet API encapsulates. For example, for the mapped drive path "Z:\\spam\\eggs", it returns a UNC path of the form "\\\\server\\share\\some\\filesystem\\path\\spam\\eggs".
* A join that tries to resolve ".." against the drive or share root path must fail, whereas this is ignored for the root path in POSIX. For example, `symlink_join("C:\\", "..\\spam")` must fail, since the system would fail an open that tried to reparse that symlink target.
* At the end, the resolved path should be tested to try to remove "\\\\?\\" if the source path didn't have this prefix. Call `GetFullPathNameW()` to check for a reserved name in the final component and `PathCchCanonicalizeEx()` to check for long-path support. (The latter calls the system runtime library function `RtlAreLongPathsEnabled`, but that's an undocumented implementation detail.)

`GetFinalPathNameByHandleW()` is not required. Optionally, it can be called for the last valid component if the caller wants a final path with all mountpoints resolved, i.e. add a `final_path=False` option. Of course, a final UNC path must retain mountpoints, so there's nothing we can do in that case. It's fine that this `realpath()` implementation would return a path that contains mountpoints in Windows (as the current implementation also does for UNC paths). They are not symlinks, and this matches the behavior of POSIX.

I'd include a warning in the documentation that getting a final path via `GetFinalPathNameByHandleW()` in the non-strict case may be dysfunctional. The unresolved tail end of the path may become valid again if a server or device comes back online. If the unresolved part contains symlinks with relative targets such as "\\spam" and "..\\..\\spam", and the `realpath()` call resolved away mountpoints, the reminaing path may not resolve correctly against the final path, as compared to how it would resolve against the original path. It definitely will not resolve the same for a rooted target path such as "\\spam" if the last resolved reparse point in the original path was a mountpoint, since it will reparse to the root path of the mountpoint device instead of the original opened device, or instead of the last resolved device of a symlink in the path.

barneygale replied:
Thanks for the detailed explanation, that's fantastic.

Are you happy for me to log a bug and include your write-up? Even with the caveats/incompatibility you mention, your proposal still seems like a substantial improvement.

barneygale replied:
I've logged a follow-up bug here so this doesn't get lost: https://bugs.python.org/issue43936

barneygale committed:
Fix indentation

barneygale committed:
Copy description of 'strict' from pathlib.

barneygale committed:
Add tests

barneygale committed:
Add NEWS

barneygale committed:
Fix ntpath tests (pass 1)

zooba commented on Doc/library/os.path.rst:
```diff
@@ -344,22 +344,29 @@ the :mod:`glob` module.)
-      When symbolic link cycles occur, the returned path will be one member of
-      the cycle, but no guarantee is made about which member that will be.
+   If the path doesn't exist and *strict* is ``True``, :exc:`FileNotFoundError`
+   is raised. If *strict* is ``False``, the path is resolved as far as possible
```
This means the target path, right? If `path` doesn't exist there'll always be an exception regardless of `strict`.

barneygale replied:
In non-strict mode, `os.path.realpath()` doesn't need _any_ part of the path to exist at _any_ stage. This doesn't just apply to broken symlink targets.

```python
>>> os.path.realpath('/foo/bar/baz')
'/foo/bar/baz'
```

barneygale replied:
FYI this is a straight copy-paste of the description of `strict` for `pathlib.Path.resolve()`

zooba commented on Misc/NEWS.d/next/Library/2021-04-08-22-11-27.bpo-25264.hEqA5d.rst:
```diff
@@ -0,0 +1,6 @@
+:func:`os.path.realpath` now raises :exc:`OSError` when a symlink loop is
+encountered. Previously it returned a path with unresolved symlinks.
+
+:func:`os.path.realpath` now accepts a *strict* keyword-only argument.
```
I think this needs to be a single paragraph. Alternatively, you can add a second NEWS file by copying the name and just changing the random letters in the last part of the filename.

zooba reviewed (commented):
Not declaring an opinion yet on the behaviour change, but I _think_ it's probably okay. I'd be happier if the link cycle change was also tied to the strict argument. (My main concern is that virtually nobody will be testing for it, and it can become an issue based entirely on user's setup - nothing to do with the code that will fail.)

barneygale committed:
Add some notes on OS differences

barneygale committed:
Split NEWS

eryksun commented on Lib/ntpath.py:
```diff
@@ -659,6 +658,8 @@ def realpath(path):
             path = _getfinalpathname(path)
             initial_winerror = 0
         except OSError as ex:
+            if strict:
```
It may as well handle `ERROR_CANT_RESOLVE_FILENAME` here instead of calling `_getfinalpathname` again.

```python
    # ERROR_CANT_RESOLVE_FILENAME (1921) is from exceeding the
    # max allowed number of reparse attempts (currently 63), which
    # is either due to a loop or a chain of links that's too long.
    if strict or ex.winerror == 1921:
        raise
```
I agree that there's no point in returning a path that has a link that can't be resolved due to a cycle or too many links. It's a bad path, and someone needs to be made aware of it by raising an exception. But I haven't read the reasons and use cases that led to the current design.

barneygale committed:
Do not suppress initial ERROR_CANT_RESOLVE_FILENAME error in non-strict mode.

barneygale committed:
realpath(): only raise OSError for symlink loops in strict mode

barneygale committed:
Fix test_ntpath (pass 1)

barneygale committed:
Fix test_ntpath (pass 2)

barneygale committed:
Split up exception handling in resolve() for greater clarity.

zooba reviewed (approved):
I'm happy with this now, just want to wait for one more approval before merging.
